spec/core/ics-003-connection-semantics/UPGRADES.md

```diff
@@ -0,0 +1,483 @@
+    abortTransactionUnless(connection.state == UPGRADE_INIT || connection.state == UPGRADE_TRY)
+
+    abortTransactionUnless(!isEmpty(errorReceipt))
+
+    abortTransactionUnless(verifyUpgradeError(currentConnection, proofHeight, proofUpgradeError, currentConnection.counterpartyConnectionIdentifier, errorReceipt))
+
+    // cancel upgrade
+    // and restore original conneciton
+    // delete unnecessary state
+    originalConnection = privateStore.get(restorePath(identifier))
+    provableStore.set(connectionPath(identifier), originalConnection)
+
+    // delete auxilliary upgrade state
+    provableStore.delete(timeoutPath(identifier))
+    privateStore.delete(restorePath(identifier))
+}
+```
+
+### Timeout Upgrade Process
+
+It is possible for the connection upgrade process to stall indefinitely on UPGRADE_TRY if the UPGRADE_TRY transaction simply cannot pass on the counterparty; for example, the upgrade feature may not be enabled on the counterparty chain.
+
+In this case, we do not want the initializing chain to be stuck indefinitely in the `UPGRADE_INIT` step. Thus, the `UpgradeInit` message will contain a `TimeoutHeight` and `TimeoutTimestamp`. The counterparty chain is expected to reject `UpgradeTry` message if the specified timeout has already elapsed.
+
+A relayer must then submit an `UpgradeTimeout` message to the initializing chain which proves that the counterparty is still in its original state. If the proof succeeds, then the initializing chain shall also restore its original connection and cancel the upgrade.
+
+```typescript
+function timeoutConnectionUpgrade(
+    identifier: Identifier,
+    counterpartyConnection: ConnectionEnd,
+    proofConnection: CommitmentProof,
+    proofHeight: Height,
+) {
+    // current connection must be in UPGRADE_INIT
+    currentConnection = provableStore.get(connectionPath(identifier))
+    abortTransactionUnles(currentConnection.state == UPGRADE_INIT)
+
+    upgradeTimeout = provableStore.get(timeoutPath(identifier))
+
+    // proof must be from a height after timeout has elapsed. Either timeoutHeight or timeoutTimestamp must be defined.
+    // if timeoutHeight is defined and proof is from before timeout height
+    // then abort transaction
+    abortTransactionUnless(upgradeTimeout.timeoutHeight.IsZero() || proofHeight >= upgradeTimeout.timeoutHeight)
+    // if timeoutTimestamp is defined then the consensus time from proof height must be greater than timeout timestamp
+    consensusState = queryConsensusState(currentConnection.clientIdentifer, proofHeight)
+    abortTransactionUnless(upgradeTimeout.timeoutTimestamp.IsZero() || consensusState.getTimestamp() >= upgradeTimeout.timestamp)
+
+    // counterparty connection must be proved to still be in OPEN state
+    abortTransactionUnless(counterpartyConnection.State === OPEN)
+    abortTransactionUnless(connection.client.verifyConnectionState(proofHeight, proofConnection, counterpartyConnection))
```

Comment:
Use the connection helper function, i.e., https://github.com/cosmos/ibc/blob/master/spec/core/ics-003-connection-semantics/README.md#helper-functions